src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+                    if ratio <= bound {
+                        return true;
+                    }
+                }
+                false
+            }
+            let mut w: f64 = u;
+            while !factorial_check(w, l) {
+                w += 1.0;
+            }
+            w
+        }
+
+        let n_p_f64 = params.n_p as f64;
+        let n_f_f64 = params.n_f as f64;
+        let lognpnf = (n_p_f64 / n_f_f64).log2();
+        let lambda_rel = params.lambda_rel as f64;
+        let logrel = lambda_rel.log2();
+        let lambda_sec = (params.lambda_sec as f64) + logrel;
+
+        let u_f64 = ((lambda_sec + logrel + 5.0 - loge.log2()) / lognpnf).ceil();
+        let u = u_f64 as usize;
+
+        let ratio = 9.0 * n_p_f64 * loge / ((17 * u).pow(2) as f64);
+        let s1 = ratio - 7.0;
+        let s2 = ratio - 2.0;
+
+        if s1 < 1.0 || s2 < 1.0 {
+            // Small case, ie n_p <= λ^2
+            let ln12 = (12f64).ln();
+            let d = (32.0 * ln12 * u_f64).ceil();
+            return Setup {
+                n_p: params.n_p,
+                u,
+                r: params.lambda_rel,
+                d: d as usize,
+                q: (2.0 * ln12 / d),
+                b: (8.0 * (u_f64 + 1.0) * d / ln12).floor() as usize,
+            };
+        }
+        let lambda_rel2 = lambda_rel.min(s2);
+        if u_f64 < lambda_rel2 {
+            // Case 3, Theorem 14, ie  n_p >= λ^3
+            let d = (16.0 * u_f64 * (lambda_rel2 + 2.0) / loge).ceil();
+            assert!(n_p_f64 >= d * d * loge / (9.0 * (lambda_rel2 + 2.0)));
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel2).ceil() as usize,
+                d: d as usize,
+                q: (2.0 * (lambda_rel2 + 2.0) / (d * loge)),
+                b: (((lambda_rel2 + 2.0 + u_f64.log2()) / (lambda_rel2 + 2.0))
+                    * (3.0 * u_f64 * d / 4.0)
+                    + d
+                    + u_f64)
+                    .floor() as usize,
+            }
+        } else {
+            // Case 2, Theorem 13, ie λ^3 > n_p > λ^2
+            let lambda_rel1 = lambda_rel.min(s1);
+            let lbar = (lambda_rel1 + 7.0) / loge;
+            let d = (16.0 * u_f64 * lbar).ceil();
+            assert!(n_p_f64 >= d * d / (9.0 * lbar));
+
+            let w = compute_w(u_f64, lambda_rel1);
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel1).ceil() as usize,
+                d: d as usize,
+                q: (2.0 * lbar / d).recip(),
+                b: (((w * lbar) / d + 1.0)
+                    * E.powf(2.0 * u_f64 * w * lbar / n_p_f64 + 7.0 * u_f64 / w)
+                    * d
+                    * u_f64
+                    + d)
+                    .floor() as usize,
+            }
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
```

Comment:
Idem: for this structure (see Setup)

Reply:
See issue https://github.com/cardano-scaling/alba/issues/38